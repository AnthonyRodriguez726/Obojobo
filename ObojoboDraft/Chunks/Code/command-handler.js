let CommandHandler
let { Editor } = window
import Common from 'Common'

let { TextGroupCommandHandler } = Editor.chunk.textChunk
let { Chunk } = Common.models
let { TextGroupSelection } = Common.textGroup

export default (CommandHandler = class CommandHandler extends TextGroupCommandHandler {
	onEnter(selection, chunk, shiftKey) {
		chunk.markDirty()

		if (!shiftKey) {
			chunk.splitText()
			return
		}

		let newChunk = Chunk.create()
		chunk.addChildAfter(newChunk)
		return newChunk.selectStart()
	}

	onSelectAll(selection, chunk) {
		chunk.selectAll()
		return true
	}

	deleteText(selection, chunk, deleteForwards) {
		chunk.markDirty()

		let tgs = new TextGroupSelection(chunk, selection.virtual)
		let data = chunk.modelState

		let s = tgs.start

		if (s.isTextStart && s.textGroupItem.data.indent > 0) {
			s.textGroupItem.data.indent--
			return
		}

		return super.deleteText(selection, chunk, deleteForwards)
	}

	// onTab: (selection, chunk, untab) ->
	// 	chunk.insertText "\t"

	indent(selection, chunk, decreaseIndent) {
		chunk.markDirty()

		let data = chunk.modelState
		let tgs = new TextGroupSelection(chunk, selection.virtual)

		let all = tgs.getAllSelectedTexts()

		return (() => {
			let result = []
			for (let textItem of Array.from(all)) {
				let item
				if (textItem.data.indent != null && !isNaN(textItem.data.indent)) {
					if (!decreaseIndent) {
						item = textItem.data.indent++
					} else if (textItem.data.indent > 0) {
						item = textItem.data.indent--
					}
				}
				result.push(item)
			}
			return result
		})()
	}
})
