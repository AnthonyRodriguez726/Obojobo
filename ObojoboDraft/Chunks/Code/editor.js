import React from 'react'

const CODE_NODE = 'ObojoboDraft.Chunks.Code'
const TEXT_NODE = 'ObojoboDraft.Chunks.Text'
const CODE_LINE_NODE = 'ObojoboDraft.Chunks.Code.CodeLine'

const Line = props => {

	return (
		<span
			className={'text align-left'}
			{...props.attributes}
			data-indent={props.node.data.get('content').indent}>
			{props.children}
		</span>
	)
}

const Node = props => {
	return (
		<div className={'component'}>
			<div
				className={`text-chunk obojobo-draft--chunks--code pad`}
				{...props.attributes}>
				<pre>
					<code>
						{props.children}
					</code>
				</pre>
			</div>
		</div>
	)
}

const isType = change => {
	return change.value.blocks.some(block => {
		return !!change.value.document.getClosest(block.key, parent => {
			return parent.type === CODE_NODE
		})
	})
}

const toggleNode = change => {
	// See if any of the selected nodes have a CODE_NODE parent
	const isCode = isType(change)

	if (isCode){
		change
			.unwrapBlock(CODE_NODE)
			.setBlocks(TEXT_NODE)
	} else {
		change
			.setBlocks(CODE_LINE_NODE)
			.wrapBlock(CODE_NODE)
	}
}

const slateToObo = node => {
	if (node.type !== CODE_NODE) return null

	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = node.data.get('content') || {}

	json.content.textGroup = []
	node.nodes.forEach(line => {
		const codeLine = {
			text: { value: line.text, styleList: [] },
			data: { indent: line.data.get('content').indent }
		}

		let currIndex = 0

		line.nodes.forEach(text => {
			text.leaves.forEach(textRange => {
				textRange.marks.forEach(mark => {
					const style = {
						start: currIndex,
						end: currIndex + textRange.text.length,
						type: mark.type,
						data: JSON.parse(JSON.stringify(mark.data))
					}
					codeLine.text.styleList.push(style)
				})
				currIndex += textRange.text.length
			})
		})

		json.content.textGroup.push(codeLine)
	})
	json.children = []

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: {} }

	json.nodes = []
	node.content.textGroup.forEach(line => {
		const indent = line.data ? line.data.indent : 0
		const codeLine = {
			object: 'block',
			type: CODE_LINE_NODE,
			data: { content: { indent: indent }},
			nodes: [
				{
					object: 'text',
					leaves: [
						{
							text: line.text.value
						}
					]
				}
			]
		}

		json.nodes.push(codeLine)
	})

	return json
}

const plugins = {
	// Basic functions, will probably want to replace with slate-edit-code
	onKeyDown(event, change) {
		// See if any of the selected nodes have a CODE_NODE parent
		const isCode = isType(change)

		// Enter
		if (isCode && event.key === 'Enter') {
			event.preventDefault()
			change.insertBlock({ type: CODE_LINE_NODE, data: { content: { indent: 0 }}})
			return true
		}

		// Shift Tab
		if (isCode && event.key === 'Tab' && event.shiftKey) {
			event.preventDefault()
			change.value.blocks.forEach(block => {
				let newIndent = block.data.get('content').indent - 1
				if(newIndent < 1) newIndent = 0

				return change.setNodeByKey(block.key, {
					data: { content: { indent: newIndent }}
				})
			})
			return true
		}

		// Tab indent
		if (isCode && event.key === 'Tab') {
			event.preventDefault()
			change.value.blocks.forEach(block => change.setNodeByKey(block.key, {
				data: { content: { indent: block.data.get('content').indent + 1 }}
			}))
			return true
		}
	},
	renderNode(props) {
		switch (props.node.type) {
			case CODE_NODE:
				return <Node {...props} />
			case CODE_LINE_NODE:
				return <Line {...props} />
		}
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.Code': {
				nodes: [{ types: [CODE_LINE_NODE] }]
			},
			'ObojoboDraft.Chunks.Code.CodeLine': {
				nodes: [{ objects: ['text'] }]
			}
		}
	}
}

const Code = {
	components: {
		Node,
		Line
	},
	helpers: {
		isType,
		toggleNode,
		slateToObo,
		oboToSlate,
	},
	plugins
}

export default Code
