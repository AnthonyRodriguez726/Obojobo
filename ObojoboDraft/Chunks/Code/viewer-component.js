import './viewer-component.scss'

import Common from 'Common'

let { OboComponent } = Common.components
let { TextGroupEl } = Common.chunk.textChunk
let { TextChunk } = Common.chunk

export default props => (
	<OboComponent model={props.model} moduleData={props.moduleData}>
		<TextChunk className="obojobo-draft--chunks--single-text pad">
			<pre>
				<code>
					{props.model.modelState.textGroup.items.map((textItem, index) => (
						<TextGroupEl
							parentModel={props.model}
							textItem={textItem}
							groupIndex={index}
							key={index}
						/>
					))}
				</code>
			</pre>
		</TextChunk>
	</OboComponent>
)
