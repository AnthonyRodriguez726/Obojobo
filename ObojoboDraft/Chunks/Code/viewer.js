import Common from 'Common'

import ViewerComponent from './viewer-component'

let SelectionHandler = Common.chunk.textChunk.TextGroupSelectionHandler
let adapter = Common.chunk.textChunk.TextGroupAdapter

Common.Store.registerModel('ObojoboDraft.Chunks.Code', {
	type: 'chunk',
	default: true,
	adapter: adapter,
	componentClass: ViewerComponent,
	selectionHandler: new SelectionHandler()
})
