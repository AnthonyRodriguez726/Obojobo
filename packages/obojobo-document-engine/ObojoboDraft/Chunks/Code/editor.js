import React from 'react'
import { Block } from 'slate'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'
import TextUtil from '../../../src/scripts/oboeditor/util/text-util'

const CODE_NODE = 'ObojoboDraft.Chunks.Code'
const CODE_LINE_NODE = 'ObojoboDraft.Chunks.Code.CodeLine'

const Line = props => {
	return (
		<span className={'text align-left'} data-indent={props.node.data.get('content').indent}>
			{props.children}
		</span>
	)
}

const Node = props => {
	return (
		<div className={'component'}>
			<div className={`text-chunk obojobo-draft--chunks--code pad`}>
				<pre>
					<code>{props.children}</code>
				</pre>
			</div>
		</div>
	)
}

const isType = change => {
	return change.value.blocks.some(block => {
		return !!change.value.document.getClosest(block.key, parent => {
			return parent.type === CODE_NODE
		})
	})
}

const insertNode = change => {
	change.insertBlock(CODE_NODE).focus()
}

const slateToObo = node => {
	const json = {}
	json.id = node.key
	json.type = node.type
	json.content = {}

	json.content.textGroup = []
	node.nodes.forEach(line => {
		const codeLine = {
			text: { value: line.text, styleList: [] },
			data: { indent: line.data.get('content').indent }
		}

		line.nodes.forEach(text => {
			TextUtil.slateToOboText(text, codeLine)
		})

		json.content.textGroup.push(codeLine)
	})
	json.children = []

	return json
}

const oboToSlate = node => {
	const json = {}
	json.object = 'block'
	json.key = node.id
	json.type = node.type
	json.data = { content: {} }

	json.nodes = []
	node.content.textGroup.forEach(line => {
		const indent = line.data ? line.data.indent : 0
		const codeLine = {
			object: 'block',
			type: CODE_LINE_NODE,
			data: { content: { indent } },
			nodes: [
				{
					object: 'text',
					leaves: TextUtil.parseMarkings(line)
				}
			]
		}

		json.nodes.push(codeLine)
	})

	return json
}

const plugins = {
	// Basic functions, will probably want to replace with slate-edit-code
	onKeyDown(event, change) {
		// See if any of the selected nodes have a CODE_NODE parent
		const isCode = isType(change)
		if (!isCode) return

		// Delete empty code node
		if (event.key === 'Backspace' || event.key === 'Delete') {
			const last = change.value.endBlock
			let parent
			change.value.blocks.forEach(block => {
				parent = change.value.document.getClosest(block.key, parent => {
					return parent.type === CODE_NODE
				})
			})

			if (last.text === '' && parent.nodes.size === 1) {
				event.preventDefault()
				change.removeNodeByKey(parent.key)
				return true
			}
		}

		// Enter
		if (event.key === 'Enter') {
			event.preventDefault()
			change.insertBlock({ type: CODE_LINE_NODE, data: { content: { indent: 0 } } })
			return true
		}

		// Shift Tab
		if (event.key === 'Tab' && event.shiftKey) {
			event.preventDefault()
			change.value.blocks.forEach(block => {
				let newIndent = block.data.get('content').indent - 1
				if (newIndent < 1) newIndent = 0

				return change.setNodeByKey(block.key, {
					data: { content: { indent: newIndent } }
				})
			})
			return true
		}

		// Tab indent
		if (event.key === 'Tab') {
			event.preventDefault()
			change.value.blocks.forEach(block =>
				change.setNodeByKey(block.key, {
					data: { content: { indent: block.data.get('content').indent + 1 } }
				})
			)
			return true
		}
	},
	renderNode(props) {
		switch (props.node.type) {
			case CODE_NODE:
				return <Node {...props} {...props.attributes} />
			case CODE_LINE_NODE:
				return <Line {...props} {...props.attributes} />
		}
	},
	renderPlaceholder(props) {
		const { node } = props
		if (node.object !== 'block' || node.type !== CODE_LINE_NODE) return
		if (node.text !== '') return

		return (
			<span className={'placeholder'} contentEditable={false}>
				{'Type Your Code Here'}
			</span>
		)
	},
	schema: {
		blocks: {
			'ObojoboDraft.Chunks.Code': {
				nodes: [
					{
						match: [{ type: CODE_LINE_NODE }],
						min: 1
					}
				],
				normalize: (change, error) => {
					const { node, child, index } = error
					switch (error.code) {
						case CHILD_TYPE_INVALID: {
							// Allow inserting of new nodes by unwrapping unexpected blocks at end and beginning
							const isAtEdge = index === node.nodes.size - 1 || index === 0
							if (child.object === 'block' && isAtEdge) {
								return change.unwrapNodeByKey(child.key)
							}

							return change.wrapBlockByKey(child.key, {
								type: CODE_LINE_NODE,
								data: { content: { indent: 0 } }
							})
						}
						case CHILD_REQUIRED: {
							const block = Block.create({
								type: CODE_LINE_NODE,
								data: { content: { indent: 0 } }
							})
							return change.insertNodeByKey(node.key, index, block)
						}
					}
				}
			},
			'ObojoboDraft.Chunks.Code.CodeLine': {
				nodes: [
					{
						match: [{ object: 'text' }],
						min: 1
					}
				],
				normalize: (change, violation, { node, child, index }) => {
					switch (violation) {
						case CHILD_TYPE_INVALID: {
							// Allow inserting of new nodes by unwrapping unexpected blocks at end and beginning
							const isAtEdge = index === node.nodes.size - 1 || index === 0
							if (child.object === 'block' && isAtEdge) {
								return change.unwrapNodeByKey(child.key)
							}
						}
					}
				}
			}
		}
	}
}

const Code = {
	components: {
		Node,
		Line
	},
	helpers: {
		insertNode,
		slateToObo,
		oboToSlate
	},
	plugins
}

export default Code
