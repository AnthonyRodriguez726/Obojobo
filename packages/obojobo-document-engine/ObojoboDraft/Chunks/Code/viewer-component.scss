@import '~styles/includes';

.obojobo-draft--chunks--code {
	pre {
		font-size: 0.9em;
		margin-top: 0;
		margin-bottom: 0;
		overflow-x: auto;
		background: $color-bg2;
		border-radius: $dimension-rounded-radius;
		box-sizing: border-box;
		padding: $dimension-padding / 4 $dimension-padding / 3 / 0.9;
		tab-size: 2;
	}

	.text {
		display: block;
		min-height: 1em;
	}
}
