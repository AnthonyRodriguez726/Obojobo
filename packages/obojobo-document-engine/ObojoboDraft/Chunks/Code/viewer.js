import Common from 'Common'
import ViewerComponent from './viewer-component'

Common.Registry.registerModel('ObojoboDraft.Chunks.Code', {
	adapter: Common.chunk.textChunk.TextGroupAdapter,
	componentClass: ViewerComponent,
	default: true,
	type: 'chunk'
})
