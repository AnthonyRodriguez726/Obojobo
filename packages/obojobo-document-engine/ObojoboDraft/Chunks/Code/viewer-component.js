import './viewer-component.scss'

import React from 'react'

import Common from 'Common'
import Viewer from 'Viewer'

const { OboComponent } = Viewer.components
const { TextGroupEl } = Common.chunk.textChunk
const { TextChunk } = Common.chunk

const Code = props => (
	<OboComponent model={props.model} moduleData={props.moduleData}>
		<TextChunk className="obojobo-draft--chunks--code pad">
			<pre>
				<code>
					{props.model.modelState.textGroup.items.map((textItem, index) => (
						<TextGroupEl
							parentModel={props.model}
							textItem={textItem}
							groupIndex={index}
							key={index}
						/>
					))}
				</code>
			</pre>
		</TextChunk>
	</OboComponent>
)

export default Code
