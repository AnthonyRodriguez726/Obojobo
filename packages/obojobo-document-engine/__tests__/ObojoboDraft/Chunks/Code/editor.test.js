import React from 'react'
import renderer from 'react-test-renderer'
import { CHILD_REQUIRED, CHILD_TYPE_INVALID } from 'slate-schema-violations'

jest.mock('../../../../src/scripts/oboeditor/util/text-util')

import Code from '../../../../ObojoboDraft/Chunks/Code/editor'
const CODE_NODE = 'ObojoboDraft.Chunks.Code'
const CODE_LINE_NODE = 'ObojoboDraft.Chunks.Code.CodeLine'

describe('Code editor', () => {
	test('Node builds the expected component', () => {
		const Node = Code.components.Node
		const component = renderer.create(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('Line builds the expected component', () => {
		const Node = Code.components.Line
		const component = renderer.create(
			<Node
				node={{
					data: {
						get: () => {
							return {}
						}
					}
				}}
			/>
		)
		const tree = component.toJSON()

		expect(tree).toMatchSnapshot()
	})

	test('insertNode calls change methods', () => {
		const change = {}
		change.insertBlock = jest.fn().mockReturnValueOnce(change)
		change.moveToStartOfNextText = jest.fn().mockReturnValueOnce(change)
		change.focus = jest.fn().mockReturnValueOnce(change)

		Code.helpers.insertNode(change)

		expect(change.insertBlock).toHaveBeenCalled()
		expect(change.focus).toHaveBeenCalled()
	})

	test('slateToObo converts a Slate node to an OboNode with content', () => {
		const slateNode = {
			key: 'mockKey',
			type: 'mockType',
			data: {
				get: () => {
					return null
				}
			},
			nodes: [
				{
					text: 'mockCode',
					data: {
						get: () => {
							return {}
						}
					},
					nodes: [
						{
							leaves: [
								{
									text: 'mockCode',
									marks: [
										{
											type: 'b',
											data: {}
										}
									]
								}
							]
						}
					]
				}
			]
		}
		const oboNode = Code.helpers.slateToObo(slateNode)

		expect(oboNode).toMatchSnapshot()
	})

	test('oboToSlate converts an OboNode to a Slate node', () => {
		const oboNode = {
			id: 'mockKey',
			type: 'mockType',
			content: {
				textGroup: [
					{
						data: { indent: 1 },
						text: { value: 'mockCode' }
					},
					{
						text: { value: 'mockCode2' }
					}
				]
			}
		}
		const slateNode = Code.helpers.oboToSlate(oboNode)

		expect(slateNode).toMatchSnapshot()
	})

	test('plugins.renderNode renders code when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: CODE_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Code.plugins.renderNode(props)).toMatchSnapshot()
	})

	test('plugins.renderNode renders a line when passed', () => {
		const props = {
			attributes: { dummy: 'dummyData' },
			node: {
				type: CODE_LINE_NODE,
				data: {
					get: () => {
						return {}
					}
				}
			}
		}

		expect(Code.plugins.renderNode(props)).toMatchSnapshot()
	})

	test('plugins.renderPlaceholder exits when not relevent', () => {
		expect(
			Code.plugins.renderPlaceholder({
				node: {
					object: 'text'
				}
			})
		).toMatchSnapshot()

		expect(
			Code.plugins.renderPlaceholder({
				node: {
					object: 'block',
					type: 'mockType'
				}
			})
		).toMatchSnapshot()

		expect(
			Code.plugins.renderPlaceholder({
				node: {
					object: 'block',
					type: CODE_LINE_NODE,
					text: 'Some text'
				}
			})
		).toMatchSnapshot()
	})

	test('plugins.renderPlaceholder renders a placeholder', () => {
		expect(
			Code.plugins.renderPlaceholder({
				node: {
					object: 'block',
					type: CODE_LINE_NODE,
					text: ''
				}
			})
		).toMatchSnapshot()
	})

	test('plugins.onKeyDown deals with no code', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => false
				},
				endBlock: {
					key: 'mockKey',
					text: 'mockText'
				}
			}
		}
		change.insertBlock = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'Enter',
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)

		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Backspace] or [Delete]', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (num, funct) => {
						funct({ key: 'mockKey' })
						return {
							key: 'mockParent',
							nodes: { size: 1 }
						}
					}
				},
				endBlock: {
					key: 'mockKey',
					text: 'mockText'
				}
			}
		}

		const event = {
			key: 'Delete',
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)
		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Backspace] or [Delete] on empty code', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: (num, funct) => {
						funct({ key: 'mockKey' })
						return {
							key: 'mockParent',
							nodes: { size: 1 }
						}
					}
				},
				endBlock: {
					key: 'mockKey',
					text: ''
				}
			}
		}
		change.removeNodeByKey = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'Delete',
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)

		expect(event.preventDefault).toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Enter]', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey'
					}
				],
				document: {
					getClosest: () => true
				}
			}
		}
		change.insertBlock = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'Enter',
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)

		expect(change.insertBlock).not.toHaveBeenCalled()
		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Shift]+[Tab]', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey',
						data: {
							get: () => {
								return { indent: 0 }
							}
						}
					}
				],
				document: {
					getClosest: () => true
				}
			}
		}
		change.setNodeByKey = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'Tab',
			shiftKey: true,
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)

		expect(change.setNodeByKey).toHaveBeenCalled()
		expect(event.preventDefault).toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Shift]+[Tab] with indented code', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey',
						data: {
							get: () => {
								return { indent: 6 }
							}
						}
					}
				],
				document: {
					getClosest: () => true
				}
			}
		}
		change.setNodeByKey = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'Tab',
			shiftKey: true,
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)

		expect(change.setNodeByKey).toHaveBeenCalled()
		expect(event.preventDefault).toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with [Tab]', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey',
						data: {
							get: () => {
								return { indent: 0 }
							}
						}
					}
				],
				document: {
					getClosest: () => true
				}
			}
		}
		change.setNodeByKey = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'Tab',
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)

		expect(change.setNodeByKey).toHaveBeenCalled()
		expect(event.preventDefault).toHaveBeenCalled()
	})

	test('plugins.onKeyDown deals with random keys', () => {
		const change = {
			value: {
				blocks: [
					{
						key: 'mockBlockKey',
						data: {
							get: () => {
								return { indent: 0 }
							}
						}
					}
				],
				document: {
					getClosest: (key, funct) => {
						funct({ type: 'mockType' })
						return true
					}
				}
			}
		}
		change.setNodeByKey = jest.fn().mockReturnValueOnce(change)

		const event = {
			key: 'e',
			preventDefault: jest.fn()
		}

		Code.plugins.onKeyDown(event, change)

		expect(change.setNodeByKey).not.toHaveBeenCalled()
		expect(event.preventDefault).not.toHaveBeenCalled()
	})

	test('plugins.schema.normalize fixes invalid children in code', () => {
		const change = {
			wrapBlockByKey: jest.fn()
		}

		Code.plugins.schema.blocks[CODE_NODE].normalize(change, {
			code: CHILD_TYPE_INVALID,
			node: { nodes: { size: 5 } },
			child: { key: 'mockKey' },
			index: null
		})

		expect(change.wrapBlockByKey).toHaveBeenCalled()
	})

	test('plugins.schema.normalize fixes invalid last block in code', () => {
		const change = {
			unwrapNodeByKey: jest.fn()
		}

		Code.plugins.schema.blocks[CODE_NODE].normalize(change, {
			code: CHILD_TYPE_INVALID,
			node: { nodes: { size: 10 } },
			child: { object: 'block', key: 'mockKey' },
			index: 0
		})

		expect(change.unwrapNodeByKey).toHaveBeenCalled()
	})

	test('plugins.schema.normalize fixes required children in code', () => {
		const change = {
			insertNodeByKey: jest.fn()
		}

		Code.plugins.schema.blocks[CODE_NODE].normalize(change, {
			code: CHILD_REQUIRED,
			node: { key: 'mockKey' },
			child: null,
			index: 0
		})

		expect(change.insertNodeByKey).toHaveBeenCalled()
	})

	test('plugins.schema.normalize does nothing to invalid children in code line', () => {
		const change = {
			unwrapBlockByKey: jest.fn()
		}

		Code.plugins.schema.blocks[CODE_LINE_NODE].normalize(change, CHILD_TYPE_INVALID, {
			node: { nodes: { size: 5 } },
			child: { key: 'mockKey' },
			index: null
		})

		expect(change.unwrapBlockByKey).not.toHaveBeenCalled()
	})

	test('plugins.schema.normalize fixes invalid last block in code line', () => {
		const change = {
			unwrapNodeByKey: jest.fn()
		}

		Code.plugins.schema.blocks[CODE_LINE_NODE].normalize(change, CHILD_TYPE_INVALID, {
			node: { nodes: { size: 10 } },
			child: { object: 'block', key: 'mockKey' },
			index: 0
		})

		expect(change.unwrapNodeByKey).toHaveBeenCalled()
	})
})
