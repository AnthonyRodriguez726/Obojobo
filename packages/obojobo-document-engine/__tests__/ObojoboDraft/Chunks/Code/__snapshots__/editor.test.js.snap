// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code editor Line builds the expected component 1`] = `
<span
  className="text align-left"
/>
`;

exports[`Code editor Node builds the expected component 1`] = `
<div
  className="text-chunk obojobo-draft--chunks--code pad"
>
  <pre>
    <code />
  </pre>
</div>
`;

exports[`Code editor oboToSlate converts an OboNode to a Slate node 1`] = `
Object {
  "data": Object {
    "content": Object {},
  },
  "key": "mockKey",
  "nodes": Array [
    Object {
      "data": Object {
        "content": Object {
          "indent": 1,
        },
      },
      "nodes": Array [
        Object {
          "leaves": undefined,
          "object": "text",
        },
      ],
      "object": "block",
      "type": "ObojoboDraft.Chunks.Code.CodeLine",
    },
    Object {
      "data": Object {
        "content": Object {
          "indent": 0,
        },
      },
      "nodes": Array [
        Object {
          "leaves": undefined,
          "object": "text",
        },
      ],
      "object": "block",
      "type": "ObojoboDraft.Chunks.Code.CodeLine",
    },
  ],
  "object": "block",
  "type": "mockType",
}
`;

exports[`Code editor plugins.renderNode renders a line when passed 1`] = `
<Line
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.Code.CodeLine",
    }
  }
/>
`;

exports[`Code editor plugins.renderNode renders code when passed 1`] = `
<Node
  attributes={
    Object {
      "dummy": "dummyData",
    }
  }
  dummy="dummyData"
  node={
    Object {
      "data": Object {
        "get": [Function],
      },
      "type": "ObojoboDraft.Chunks.Code",
    }
  }
/>
`;

exports[`Code editor plugins.renderPlaceholder exits when not relevent 1`] = `undefined`;

exports[`Code editor plugins.renderPlaceholder exits when not relevent 2`] = `undefined`;

exports[`Code editor plugins.renderPlaceholder exits when not relevent 3`] = `undefined`;

exports[`Code editor plugins.renderPlaceholder renders a placeholder 1`] = `
<span
  className="placeholder"
  contentEditable={false}
>
  Type Your Code Here
</span>
`;

exports[`Code editor slateToObo converts a Slate node to an OboNode with content 1`] = `
Object {
  "children": Array [],
  "content": Object {
    "textGroup": Array [
      Object {
        "data": Object {
          "indent": undefined,
        },
        "text": Object {
          "styleList": Array [],
          "value": "mockCode",
        },
      },
    ],
  },
  "id": "mockKey",
  "type": "mockType",
}
`;
